File: Misc/99_lc_714.java
package Misc;

/*
 * 714. Best Time to Buy and Sell Stock with Transaction Fee
 */

import java.util.*;

class Solution {
    private int helper(int startingIndex, int isBought, int[] prices, int fee, int[][] dp, int pricesLength) {
        if (startingIndex == pricesLength) {
            return 0;
        }

        if (dp[startingIndex][isBought] != -1) {
            return dp[startingIndex][isBought];
        }
        
        int profit = 0;
        if (isBought == 0) {
            profit = Math.max(helper(startingIndex+1, 0, prices, fee, dp, pricesLength), helper(startingIndex+1, 1, prices, fee, dp, pricesLength) - prices[startingIndex]);
        }else {
            profit = Math.max(helper(startingIndex+1, 1, prices, fee, dp, pricesLength), helper(startingIndex+1, 0, prices, fee, dp, pricesLength) + prices[startingIndex] - fee);
        }

        dp[startingIndex][isBought] = profit;
        return dp[startingIndex][isBought];
    }

    public int maxProfit(int[] prices, int fee) {
        int pricesLength = prices.length;
        int[][] dp = new int[pricesLength][2];

        for (int[] row: dp) {
            Arrays.fill(row, -1);
        }

        int res = helper(0, 0, prices, fee, dp, pricesLength);
        return res;
    }
}

File: Misc/99_lc_714.py
"""
714. Best Time to Buy and Sell Stock with Transaction Fee
"""

from typing import List


class Solution:
    def maxProfit(self, prices: List[int], fee: int) -> int:  # memoization
        n = len(prices)
        dp = [[-1 for _ in range(2)] for _ in range(n + 1)]

        def helper(ind, isBought):
            if ind == n:
                return 0
            if dp[ind][isBought] != -1:
                return dp[ind][isBought]

            profit = 0
            if not isBought:
                profit = max(
                    helper(ind + 1, False), helper(ind + 1, True) - prices[ind]
                )
            else:
                profit = max(
                    helper(ind + 1, True), helper(ind + 1, False) + (prices[ind] - fee)
                )

            dp[ind][isBought] = profit
            return dp[ind][isBought]

        helper(0, False)
        return dp[0][0]

    def maxProfit2(self, prices: List[int], fee: int):  # tablulation
        n = len(prices)
        dp = [[0 for _ in range(2)] for _ in range(n + 1)]

        for ind in range(n - 1, -1, -1):
            for buySell in range(2):
                profit = 0
                if buySell == 0:
                    profit = max(dp[ind + 1][0], dp[ind + 1][1] - prices[ind])
                else:
                    profit = max(dp[ind + 1][1], dp[ind + 1][0] + prices[ind] - fee)
                dp[ind][buySell] = profit

        return dp[0][0]
